File: app/src/main/java/com/dragonpav/dengine/Texture.java
package com.dragonpav.dengine;

import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.opengl.GLES30;
import android.opengl.GLUtils;
import android.util.Log;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;

public class Texture {
	public int texSampleNum = -1;
	public int textureId;
	public int glType = GLES30.GL_TEXTURE0;
	public Texture(byte[] rgb, Program p, Config cfg) {
		texSampleNum = GLES30.glGetUniformLocation(p.programId, "textureUnit");
		int num = 256* 256 * 3;
		byte[] colors = new byte[num];
		int i = 0;
		ByteBuffer colorBuffer = ByteBuffer.allocateDirect(num).order(ByteOrder.nativeOrder());
		while (i != num) {
			colorBuffer.put((byte) (rgb[0] & 0xFF));
			colorBuffer.put((byte) (rgb[1] & 0xFF));
			colorBuffer.put((byte) (rgb[2] & 0xFF));
			i += 3;
		}
		colorBuffer.position(0);
		int[] tmpBuf = new int[1];
		GLES30.glGenTextures(1, tmpBuf, 0);
		textureId = tmpBuf[0];
		GLES30.glBindTexture(GLES30.GL_TEXTURE_2D, textureId);
		GLES30.glTexParameteri(GLES30.GL_TEXTURE_2D, GLES30.GL_TEXTURE_WRAP_S, cfg.textureWrapS);
		GLES30.glTexParameteri(GLES30.GL_TEXTURE_2D, GLES30.GL_TEXTURE_WRAP_T, cfg.textureWrapT);
		GLES30.glTexParameteri(GLES30.GL_TEXTURE_2D, GLES30.GL_TEXTURE_MIN_FILTER, cfg.textureMinFilter);
		GLES30.glTexParameteri(GLES30.GL_TEXTURE_2D, GLES30.GL_TEXTURE_MAG_FILTER, cfg.textureMagFilter);
		GLES30.glTexImage2D(GLES30.GL_TEXTURE_2D, 0, GLES30.GL_RGB, 256, 256, 0, GLES30.GL_RGB, GLES30.GL_UNSIGNED_BYTE, colorBuffer);
		if (cfg.useMipmap) {
			GLES30.glGenerateMipmap(GLES30.GL_TEXTURE_2D);
		}
		GLES30.glBindTexture(GLES30.GL_TEXTURE_2D, 0);
	}
	public Texture(Resources res, int resId, Program p, Config cfg) {
		BitmapFactory.Options opts = new BitmapFactory.Options();
		opts.inScaled = false;
		Bitmap img = BitmapFactory.decodeResource(res, resId, opts);
		texSampleNum = GLES30.glGetUniformLocation(p.programId, "textureUnit");
		int[] tmpBuf = new int[1];
		GLES30.glGenTextures(1, tmpBuf, 0);
		textureId = tmpBuf[0];
		GLES30.glBindTexture(GLES30.GL_TEXTURE_2D, textureId);
		GLES30.glTexParameteri(GLES30.GL_TEXTURE_2D, GLES30.GL_TEXTURE_WRAP_S, cfg.textureWrapS);
		GLES30.glTexParameteri(GLES30.GL_TEXTURE_2D, GLES30.GL_TEXTURE_WRAP_T, cfg.textureWrapT);
		GLES30.glTexParameteri(GLES30.GL_TEXTURE_2D, GLES30.GL_TEXTURE_MIN_FILTER, cfg.textureMinFilter);
		GLES30.glTexParameteri(GLES30.GL_TEXTURE_2D, GLES30.GL_TEXTURE_MAG_FILTER, cfg.textureMagFilter);
		int byteCount = img.getByteCount();
		ByteBuffer buffer = ByteBuffer.allocateDirect(byteCount).order(ByteOrder.nativeOrder());
		img.copyPixelsToBuffer(buffer);
		buffer.position(0);
		GLES30.glTexImage2D(GLES30.GL_TEXTURE_2D, 0, GLES30.GL_RGBA, img.getWidth(), img.getHeight(), 0, GLES30.GL_RGBA, GLES30.GL_UNSIGNED_BYTE, buffer);
		if (cfg.useMipmap) {
			GLES30.glGenerateMipmap(GLES30.GL_TEXTURE_2D);
		}
		GLES30.glBindTexture(GLES30.GL_TEXTURE_2D, 0);
		img.recycle();
	}
	public void dispose() {
		if (textureId != 0) {
			int[] textures = new int[] {textureId};
			GLES30.glDeleteTextures(1, textures, 0);
			textureId = 0;
		}
	}
	public static class Config {
		public boolean useMipmap = true;
		public int textureWrapS = GLES30.GL_CLAMP_TO_EDGE;
		public int textureWrapT = GLES30.GL_CLAMP_TO_EDGE;
		public int textureMinFilter = GLES30.GL_NEAREST_MIPMAP_LINEAR;
		public int textureMagFilter = GLES30.GL_NEAREST;
	}
}

File: app/src/main/cpp/Texture.cpp
#include "Texture.h"
#include <android/imagedecoder.h>
Texture::Texture(std::vector<GLubyte> rgb, Program* p, Config cfg) {
	texSampleNum = glGetUniformLocation(p->programId, "textureUnit");
	int i = 0;
	while (i < 256 * 256 * 3) {
		colorBuffer.push_back(rgb[0]);
		i++;
		colorBuffer.push_back(rgb[1]);
		i++;
		colorBuffer.push_back(rgb[2]);
		i++;
	}
	GLuint tmpBuf[1];
	glGenTextures(1, tmpBuf);
	textureId = tmpBuf[0];
	glBindTexture(GL_TEXTURE_2D, textureId);
	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, cfg.textureWrapS);
	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, cfg.textureWrapT);
	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, cfg.textureMinFilter);
	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, cfg.textureMagFilter);
	glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB, 256, 256, 0, GL_RGB, GL_UNSIGNED_BYTE, colorBuffer.data());
	if (cfg.useMipmap) {
		glGenerateMipmap(GL_TEXTURE_2D);
	}
	glBindTexture(GL_TEXTURE_2D, 0);
}
Texture::Texture(Program* p, AAsset* file, Config cfg) {
	texSampleNum = glGetUniformLocation(p->programId, "textureUnit");
    AImageDecoder* decoder = nullptr;
	int result = AImageDecoder_createFromAAsset(file, &decoder);
    if (result != ANDROID_IMAGE_DECODER_SUCCESS) {
        __android_log_write(ANDROID_LOG_ERROR, "DEngine", "Failed to decode image");
        return;
    }
    const AImageDecoderHeaderInfo* header = AImageDecoder_getHeaderInfo(decoder);
    int width = AImageDecoderHeaderInfo_getWidth(header);
    int height = AImageDecoderHeaderInfo_getHeight(header);
    int format = AImageDecoderHeaderInfo_getAndroidBitmapFormat(header);
    if (format != ANDROID_BITMAP_FORMAT_RGB_565) {
        __android_log_write(ANDROID_LOG_ERROR, "DEngine", "No such format");
        AImageDecoder_delete(decoder);
        return;
    }
	size_t size = width * height * 3;
    int stride = AImageDecoder_getMinimumStride(decoder);
    int decodeResult = AImageDecoder_decodeImage(decoder, colorBuffer.data(), stride, size);
	if (decodeResult != ANDROID_IMAGE_DECODER_SUCCESS) {
        __android_log_write(ANDROID_LOG_ERROR, "DEngine", "Failed to decode texture");
        return;
    }
	GLuint tmpBuf[1];
	glGenTextures(1, tmpBuf);
	textureId = tmpBuf[0];
	glBindTexture(GL_TEXTURE_2D, textureId);
	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, cfg.textureWrapS);
	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, cfg.textureWrapT);
	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, cfg.textureMinFilter);
	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, cfg.textureMagFilter);
	glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB, width, height, 0, GL_RGB, GL_UNSIGNED_BYTE, colorBuffer.data());
	if (glGetError() != GL_NO_ERROR) {
		std::cerr << "Texture setup failure: gl error: " << glGetError() << std::endl;
	}
	if (cfg.useMipmap) {
		glGenerateMipmap(GL_TEXTURE_2D);
	}
	glBindTexture(GL_TEXTURE_2D, 0);
    AImageDecoder_delete(decoder);
}
void Texture::dispose() {
	glDeleteTextures(1, &textureId);
}